### PR TITLE
IF operation

### LINKED ISSUE #67: Conditions in clickRoutines
1) Add special parameter "if" to operations which only executes the operation if a condition is met

```
{
  "func": "FLIP",
  "holder": "<holderID>",
  "face": 0,
   "if": "myVariableName"
}
```

Reason: Currently, the only option to make conditional operations in clickRoutines is to use the skip parameter, which prevents an operation from executing if the value evaluates to `true`. However, oftentimes the opposite is needed: An operation should only be executed if a certain condition is met. When "if" and "skip" are used in parallel, "skip" should override "if".

2) Introduce a new function JUMP which jumps to a particular routine without losing the clickRoutine's state (variables and collections). Like any other operation, it can be combined with "if" and "skip". The routine would simply be defined as a custom property of the button widget.

Probably not the most realistic example, but should help to illustrate:

```
{
  "type": "button",
  "text": "myButton",
  "clickRoutine": [
    {
        "func": "INPUT",
        "header": "A header text for the overlay.",
        "fields": [
           {
               "type": "checkbox",
               "label": "Do you want to shuffle?",
               "variable": "shuffle",
               "value": true
           }
        ]
    },
    {
      "func": "JUMP",
      "routine": "doShuffle",
      "if": "shuffle"
    },
  ],
  "doShuffle": [
    {
      "func": "SHUFFLE",
      "holder": "<holderID>",
    }
  ],
  "x": 758,
  "y": 721,
  "z": 11,
  "id": "2zye",
  "layer": 1
}

```

Reason: Conditional branching is extremely useful but quite difficult to implement. A "jump" is one of the easiest programming concepts and together with "if" and "skip" could accomplish most of what "if/else" conditions can do.

### PR DESCRIPTION
**Pull request by larkob**
This PR introduces the IF operation, which checks a `variable` that is evaluated as true or false and then branches into the corresponding `thenRoutine` or `elseRoutine` before continuing with the remaining operations in the clickRoutine.

Here's a short example in pseudo code that flips this or that widget depending on _aValue_ being equal to 1:
```
if (aValue == 1) {
 select["DEFAULT"] = widget['flipThis']
} else {
 select["DEFAULT"] = widget['flipThat']
}
flip(select["DEFAULT"])
```

The corresponding clickRoutine looks as follows:

```
"clickRoutine": [
    {
      "func": "COMPUTE",
      "applyVariables": [
        {
          "parameter": "operand1",
          "variable": "aValue"
        }],
      "operation": "==",
      "operand2": 1,
      "variable": "isOne"
    },
    {
      "func": "IF",
      "variable": "isOne",
      "thenRoutine": [
        {
          "func": "SELECT",
          "property": "id",
          "mode": "set",
          "value": "flipThis"
        }
      ],
      "elseRoutine": [
        {
          "func": "SELECT",
          "property": "id",
          "mode": "set",
          "value": "flipThat"
        }
      ]
    },
    {
      "func": "FLIP"
    }
```

Closes #67

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3365/pr-test (or any other room on that server)

**Comment by 96LawDawg** [comment]
Works as advertised.  I added a slightly more complicated example into the test room.  I think this will also work well with SKIP and CALL to create more interesting and more refined code branching opportunities.

In trying to break the new feature, I found that if you use just a "thenRoutine" and completely remove the "elseRoutine" then the room freezes up briefly if trying to use the "else" branch.  However, if you use just "else" and not "then," then nothing bad happens but even the "else" branch doesn't work.  Users shouldn't do this, but they will by mistake.  Of course, it is okay to have an "else" that is just empty curly braces.  

Speaking of making it user friendly, I hope there will be a way to incorporate this into the JSON editor.  I can approve this as is.  Just wondering about your thoughts about the one code crashing thing I found.

**Comment by ArnoldSmith86** [comment]
Wouldn't it be better to have the condition inside the `IF`? Similar to how `SELECT` does it?

It will probably also be weird to have the `variable` parameter once we get rid of `applyVariables`..

File: client/js/widgets/widget.js [hunk 1]
```diff
@@ -599,6 +601,29 @@ export class Widget extends StateManaged {
...
+            result = await this.evaluateRoutine(a.elseRoutine, inheritVariables, inheritCollections, (depth || 0) + 1);
+          }
+          const returnedVariables = Object.entries(result.variable);
+          for (const [k,v] of returnedVariables)
+            variables[k] = v;
+          const returnedCollections = Object.entries(result.collection);
+          for (const [k,v] of returnedCollections)
+            collections[k] = v;
+        } else
+          problems.push(`IF operation is missing the 'variable' or 'then' parameter.`);
```
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:625]: The parameter is called `thenRoutine`.

File: client/js/widgets/widget.js [hunk 2]
```diff
@@ -599,6 +601,29 @@ export class Widget extends StateManaged {
           }
         }
       }
+      if(a.func == 'IF') {
+        if(typeof a.variable != 'undefined' && Array.isArray(a.thenRoutine)) {
```
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:605]: Why not allow an `elseRoutine` without `thenRoutine`?

File: client/js/widgets/widget.js [hunk 3]
```diff
@@ -599,6 +601,29 @@ export class Widget extends StateManaged {
...
+          const inheritVariables = variables;
```
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:606]: The whole inheritance stuff seems too complicated. I'll play around with it when I'm at a PC but if it works like this, we can also merge it first.

**Review by ArnoldSmith86 — APPROVED** [review]
I don't see major code problems. I didn't really test anything except clicking the buttons in the test room.

File: client/js/widgets/widget.js [hunk 4]
```diff
@@ -602,27 +602,30 @@ export class Widget extends StateManaged {
         }
       }
       if(a.func == 'IF') {
-        if(typeof a.variable != 'undefined' && Array.isArray(a.thenRoutine)) {
+        setDefaults(a, { relation: '==' });
+        if (!['==', '!=', '<', '<=', '>=', '>'].indexOf(a.relation)) {
```
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:606]: Did you test this? I didn't but the condition should be `true` for `==` (index `0`) and `false` for everything else (index `-1` or positive).

**Comment by 96LawDawg** [comment]
The two examples in the test room no longer work.

**Comment by larkob** [comment]
I've changed the operation to be more consistent with SELECT and COMPUTE. Now it asks for a relation (==, >= etc.) together with two operands. 

I wanted to ensure that checking for something to be true is still easy, so instead of two comparing two operands you can alternatively check if a `condition` parameter is true.

```
IF with 2 operands:

"clickRoutine": [
    {
      "func": "IF",
      "applyVariables": [
        {
          "parameter": "operand1",
          "variable": "aValue"
        }
      ],
      "relation": "==",
      "operand2": 1,
      "thenRoutine": [
        {
          "func": "SELECT",
          "property": "id",
          "mode": "set",
          "value": "flipThis"
        }
      ],
      "elseRoutine": [
        {
          "func": "SELECT",
          "property": "id",
          "mode": "set",
          "value": "flipThat"
        }
      ]
    },
    {
      "func": "FLIP"
    }
]
```

```
IF with condition:

"clickRoutine": [
    {
      "func": "COMPUTE",
      "applyVariables": [
        {
          "parameter": "operand1",
          "variable": "aValue"
        }],
      "operation": "==",
      "operand2": 1,
      "variable": "isOne"
    },
    {
      "func": "IF",
      "applyVariables": [
        {
          "parameter": "condition",
          "variable": "isOne"
        }
      ],
      "thenRoutine": [
        {
          "func": "SELECT",
          "property": "id",
          "mode": "set",
          "value": "flipThis"
        }
      ],
      "elseRoutine": [
        {
          "func": "SELECT",
          "property": "id",
          "mode": "set",
          "value": "flipThat"
        }
      ]
    },
    {
      "func": "FLIP"
    }
]
```

My hope is that we'll eventually support use of variables and simple expressions in parameters, so that the IF routine could look sth. like this:

```
"clickRoutine": [
    {
      "func": "IF",
      "condition: "${aValue == 1}",
      "thenRoutine": [
        {
          "func": "SELECT",
          "property": "id",
          "mode": "set",
          "value": "flipThis"
        }
      ],
      "elseRoutine": [
        {
          "func": "SELECT",
          "property": "id",
          "mode": "set",
          "value": "flipThat"
        }
      ]
    },
    {
      "func": "FLIP"
    }
]
```

File: client/js/widgets/widget.js [hunk 5]
```diff
@@ -599,6 +601,33 @@ export class Widget extends StateManaged {
           }
         }
       }
+      if(a.func == 'IF') {
+        setDefaults(a, { relation: '==' });
+        if (!['==', '!=', '<', '<=', '>=', '>'].indexOf(a.relation) < 0) {
```
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:606]: Did you mean to remove the `!`? If you didn't, I really don't understand how this works. :D

Comment (larkob) [reply client/js/widgets/widget.js]: yeah, thanks

**Comment by 96LawDawg** [comment]
The IF with condition seems to work (that is how the checker scenario is set up). The card scenario uses IF with 2 operands.  That does not work.

**Comment by larkob** [comment]
@96LawDawg Can you please check again? I've just tested it and cannot find any problem.

**Comment by 96LawDawg** [comment]
It is supposed to deal 5 cards if dice roll >=4 and 1 otherwise.  It doesn't work consistently.

https://user-images.githubusercontent.com/76912527/114321643-7b301580-9ae1-11eb-9923-7413e9530ef9.mp4

**Comment by larkob** [comment]
The routine clicks the spinner after moving cards, it evens says so in the label. So, the move is based on the previous "dice roll", which is exactly what happens in the video.

**Review by 96LawDawg — APPROVED** [review]
I am truly an idiot and forgot what my own test was supposed to do.  Approved through test.

**Comment by ArnoldSmith86** [comment]
I changed the code so that the `thenRoutine` and `elseRoutine` "directly" work on the variables and collections of the parent routine.

I didn't test this extensively but the examples in the pr-test room seem to still work.

**Comment by 96LawDawg** [comment]
Something is wrong with the JSON editor in this version. Trying to select the "thenRoutine" or "elseRoutine" forces a shutdown of the JSON editor. Selecting anything else is fine.

**Comment by larkob** [comment]
It's a bug on the main branch. I'll take a look to see how it can be fixed.

**Comment by larkob** [comment]
@ArnoldSmith86 I need help with the JSON editor issue. It assumes that any Routine is on the first level of the widget, which was correct so far, but now the IF operation allows for a Routine inside a Routine. I cannot figure out how to resolve this issue. The browser crashes on the main branch as well when you add a routine inside another routine and focus on it in the JSON editor.

**Comment by robartsd** [comment]
> @ArnoldSmith86 I need help with the JSON editor issue. It assumes that any Routine is on the first level of the widget, which was correct so far...

Routines in a deck can appear in `cardDefaults` or in entries in `cardTypes` and routines for a basic widget can appear in entries `faces`. I think this was correctly fixed (at least for `cardDefaults`); but nesting routines within a routine element may be a new animal.

**Comment by ArnoldSmith86** [comment]
Yeah, this is getting complicated... :D

It's also great fun in the file updater.

I'll have a look.

**Comment by robartsd** [comment]
> It's also great fun in the file updater.

This alone might be a good enough reason for not nesting routines but calling routines from the IF instead; much simpler maintenance when things change later.

**Comment by larkob** [comment]
Yes, I did consider that, but it makes reading and understanding more difficult. How about we don't use "Routine" but call them "thenBranch" and "elseBranch"? And "loopSequence" for the upcoming FOR operation.

**Comment by ArnoldSmith86** [comment]
I'm pretty sure that only makes it worse. There can always be an `IF` inside the `thenRoutine`.

It'll be fine.

**Comment by ArnoldSmith86** [comment]
I fixed the crash in the JSON editor and I think it's working fine but there are still some issues:

1. In nested routines it will display commands for unused parameters for all active commands (so a parent `IF` and a child `SELECT` for example) but clicking them will always add the parameter to the child command. So for now you'll have to move the text cursor out of the `thenRoutine`.
2. For nested routines commands should be displayed more than once (add a `SELECT` to the parent routine or add it to the child routine). Making that work will be quite a bit of work. I'm pretty sure I want it that way but I don't think I want to implement that _now_ in this PR.

**Comment by 96LawDawg** [comment]
I tested the changes to the JSON editor and they work.  In my opinion, everything works now and this is ready to be merged.

**Pull request merged**